keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+
+  - Parallel node creation
+
+    Higher-level tools could create nodes in parallel (both masters and workers)
+    for reducing the overall cluster startup time.
+    `kubeadm join --master` should support natively this practice without requiring
+    the implementation of any synchronization mechanics by higher-level tools.
+
+  - Replace reconciliation strategies
+
+    Especially in case of cloud deployments, higher-level automation tools could
+    decide for any reason to replace existing nodes with new ones (instead of apply
+    changes in-place to existing nodes).
+    `kubeadm join --master` will support this practice by making easier to replace
+    existing master nodes with new ones.
+
+- Provide support both for dynamic and static bootstrap flow
+
+  At the time a user is running `kubeadm init`, s/he/an operator might not know what
+  the cluster setup will look like eventually. For instance, the user may start with
+  only one master + n nodes, and then add further master nodes with `kubeadm join --master`
+  or add more worker nodes with `kubeadm join` (in any order). This kind of workflow, where the
+  user doesn’t know in advance the final layout of the control plane instances, into this
+  document is referred as “dynamic bootstrap workflow”.
+
+  Nevertheless, kubeadm should support also more “static bootstrap flow”, where a user knows
+  in advance the target layout of the controlplane instances (the number, the name and the IP
+  of master nodes).
+
+### Non-goals
+
+- This proposal doesn't include a solution for etcd cluster management (but nothing in this proposal should
+  prevent to address this in future).
+
+  > At the time of writing, the CoreOS recommended approach for etcd is to run
+  > the etcd cluster outside kubernetes (see discussion in [kubeadm office hours](https://goo.gl/fjyeqo)).
+
+- Nothing in this proposal should prevent users to run master nodes components
+  and etcd on the same machines.
+
+- This proposal doesn't include a solution for API server load balancing (Nothing in this proposal
+  should prevent users from choosing their preferred solution for API server load balancing.
+
+- This proposal doesn't address the ongoing discussion about kubeadm self-hosting; in light of
+  divide and conquer goal stated before, it is not planned to provide initial support
+  for self-hosted clusters (but nothing in this proposal should prevent to address this in future).
+
+- This proposal doesn't address the ongoing discussion about storage of kubeadm TLS assets
+  in secrets; in light of divide and conquer goal stated before, it is not planned to
+  provide initial support for clusters with TLS stored in secrets (but nothing in this
+  proposal should prevent to address this in future).
+
+- Nothing in this proposal should prevent practices that exist today.
+
+### Challenges and Open Questions
+
+- Keep the UX simple.
+
+  - _What are the acceptable trade-offs between the need to have a clean and simple
```

Comment:
should you/we answer this?

Reply:
yes! in the implementation is described how we can support different types of HA clusters / join workflow by making only few changes to the `kubeadm join` UX